test/tools/cli/IonProcess.java

```diff
@@ -0,0 +1,314 @@
+            ionWriterForErrorReport.finish();
+        }
+        catch (Exception e){
+            System.err.println("Finish ionWriter Error: " + e.getMessage());
+        }
+    }
+
+    private static void processEvents(IonWriter ionWriterForOutput,
+                                     IonWriter ionWriterForErrorReport,
+                                     IonReader ionReader) throws Exception {
+        //curTable is used for checking if the symbol table changes during the processEvent function.
+        SymbolTable curTable = ionReader.getSymbolTable();
+
+        processEvent(ionWriterForOutput, ionWriterForErrorReport, ionReader, curTable);
+        new Event(EventType.STREAM_END,null,null,null,
+                null,null, null, 0 )
+                .writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+    }
+
+    private static void processEvent(IonWriter ionWriterForOutput,
+                                      IonWriter ionWriterForErrorReport,
+                                      IonReader ionReader,
+                                      SymbolTable curTable) throws Exception {
+        while(ionReader.next() != null){
+            if(!isSameSymbolTable(ionReader.getSymbolTable(), curTable)) {
+                curTable = ionReader.getSymbolTable();
+                ImportDescriptor imports[] = symbolTableToImports(curTable.getImportedTables());
+                new Event(EventType.SYMBOL_TABLE,null,null,null,
+                        null,null, imports, 0 )
+                        .writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+            }
+
+            if (IonType.isContainer(ionReader.getType())){
+                //get current Ion type and depth
+                IonType curType = ionReader.getType();
+                int curDepth = ionReader.getDepth();
+
+                //write a Container_Start event and step in
+                ionStreamToEvent(EventType.CONTAINER_START,ionReader)
+                        .writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+                ionReader.stepIn();
+
+                //recursive call
+                processEvent(ionWriterForOutput, ionWriterForErrorReport, ionReader, curTable);
+
+                //write a Container_End event and step out
+                new Event(EventType.CONTAINER_END,curType,null,null,
+                        null,null,null, curDepth )
+                        .writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+                ionReader.stepOut();
+            } else {
+                ionStreamToEvent(EventType.SCALAR,ionReader).writeOutput(ionWriterForOutput, ionWriterForErrorReport);
+            }
+        }
+    }
+
+    /**
+     * this function is only used for CONTAINER_START and SCALAR eventType since other types don't need initial data.
+     */
+    private static Event ionStreamToEvent(EventType eventType, IonReader ionReader) throws Exception {
+        if(ionReader.getType() == null) return null;
```

Comment:
It seems like this is actually an illegal state. Should you throw `IllegalStateException` in this case, or simply state that callers must provide a reader that is positioned on a value?

Reply:
I throw an IllegalStateException(IonProcess.java 200)